File: KseniaLikhosherstova/Term2/Block2/Task2/Game/Cards/Card.cs
namespace Game.Cards
{
    public class Card
    {
        public Suit Suit { get; }
        public Rank Rank { get; }


        public Card(Rank rank, Suit suit)
        {
            if (!Enum.IsDefined(rank) || !Enum.IsDefined(suit))
                throw new ArgumentException($"Rank {rank} or/and suit {suit} are not valid");

            Rank = rank;
            Suit = suit;
        }



        public int GetValueOfCard(int score)
        {
            if (Rank == Rank.King || Rank == Rank.Queen || Rank == Rank.Jack)
                return 10;

            if (Rank == Rank.Ace)
            {
                if ((score + 11) > 21)
                    return 1;
                else
                    return 11;
            }

            return (int)Rank;
        }

        public override string ToString()
        {
            int d = (int)Rank;
            string name = d switch
            {
                (<= 10) => d.ToString(),
                11 => "Jack",
                12 => "Queen",
                13 => "King",
                14 => "Ace",
                _ => "",
            };

            string suit = Suit switch
            {
                Suit.Clubs => "of clubs",
                Suit.Diamonds => "of diamonds",
                Suit.Hearts => "of hearts",
                Suit.Spades => "of spades",
                _ => "",
            };

            return $"{name} {suit}";
        }
    }
}

File: AzimMuradov/Term2/Block3/Task4/lib/base/src/main/kotlin/casino/lib/card/Card.kt
package casino.lib.card

@JvmInline
value class Card private constructor(private val value: Int) {

    val suite: CardSuite get() = CardSuite.values()[value / CardRank.values().size]

    val rank: CardRank get() = CardRank.values()[value % CardRank.values().size]


    operator fun component1(): CardSuite = suite

    operator fun component2(): CardRank = rank


    override fun toString(): String {
        val suiteCodePart = when (suite) {
            CardSuite.SPADES -> "1f0a"
            CardSuite.HEARTS -> "1f0b"
            CardSuite.DIAMONDS -> "1f0c"
            CardSuite.CLUBS -> "1f0d"
        }
        val rankCodePart = when (rank) {
            CardRank.QUEEN, CardRank.KING -> rank.ordinal + 2
            else -> rank.ordinal + 1
        }.toString(radix = 16)

        return String(Character.toChars("$suiteCodePart$rankCodePart".toInt(radix = 16)))
    }


    companion object {

        fun get(suite: CardSuite, rank: CardRank): Card = Card(
            value = suite.ordinal * CardRank.values().size + rank.ordinal
        )

        val deck: List<Card> = List(size = CardSuite.values().size * CardRank.values().size, init = ::Card)
    }
}
